File: book_kit/src/main/java/net/lzbook/kit/appender_loghub/appender/AndroidLogClient.java
package net.lzbook.kit.appender_loghub.appender;


import net.lzbook.kit.appender_loghub.LOGClient;
import net.lzbook.kit.appender_loghub.LogGroup;
import net.lzbook.kit.appender_loghub.ServerLog;
import net.lzbook.kit.utils.logger.AppLog;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;


public class AndroidLogClient {

    public static final String endPoint = "cn-shenzhen.log.aliyuncs.com";// Endpoint
    public static final String accessKeyId = "LTAIHv56dMm1Dd5Z"; // 使用您的阿里云访问密钥
    public static final String accessKeySecret = "30hIE7U1i6D4azaCwsWnFWS19G4yAb"; // 使用您的阿里云访问密钥
    //	private final static Logger logger = LoggerFactory.getLogger(AndroidLogClient.class);
    // 定义发送队列
    private static volatile Queue<LogGroup> send_queue = new ConcurrentLinkedQueue<LogGroup>();
    private static LogSendThread senderThread;

    static {
        senderThread = new LogSendThread();
        senderThread.start();
    }

    public static void putLog(List<ServerLog> logList) {
        try {
            //按照project和logstore分组，所以log中必须包含project和logstore两个key
            Map<String, LogGroup> map = new HashMap<>();
            if (logList != null && logList.size() > 0) {
                for (int i = 0; i < logList.size(); i++) {
                    if (!logList.get(i).getContent().containsKey("project") || !logList.get(i).getContent().containsKey("logstore")) {
//						loggerer.error("log not containsKey : project || logstore");
                        continue;
                    }
                    String project = (String) logList.get(i).getContent().get("project");
                    String logstore = (String) logList.get(i).getContent().get("logstore");
                    String unikey = project + "_" + logstore;
                    if (!map.containsKey(unikey)) {
                        LogGroup logGroup = new LogGroup("", "", project, logstore);
                        map.put(unikey, logGroup);
                    }
                    map.get(unikey).PutLog(logList.get(i));
                    AppLog.e("ad", (String) logList.get(i).getContent().get("project"));
                }
                if (!map.values().isEmpty()) {
                    for (LogGroup logGroup : map.values()) {
                        send_queue.add(logGroup);
                        senderThread.wakeUp();
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
//			logger.error("put element to queue error ", e);
        }
    }

    // 开启一个线程专门用来发送数据
    private static class LogSendThread extends Thread {
        private volatile boolean loop = true;
        private volatile boolean is_sleeping = false;

        public LogSendThread() {
            super();
        }

        public void run() {
            while (loop) {
                is_sleeping = false;
                if (send_queue.isEmpty()) {
                    is_sleeping = true;
                    synchronized (this) {
                        try {
                            this.wait();
                        } catch (InterruptedException e) {
                        }
                    }
                    continue;
                }
                LogGroup logGroup = send_queue.poll();
                String project = logGroup.getProject();
                String logstore = logGroup.getLogstore();
                try {
                    if (logGroup != null && logGroup != null) {
                        LOGClient logClient = new LOGClient(endPoint, accessKeyId, accessKeySecret, project);
                        logClient.PostLog(logGroup, logstore);
                    }
                    // 是否要有重试功能.......
                } catch (Exception e) {
                    e.printStackTrace();
//						logger.error("send log error", e);
                    AndroidLogStorage.getInstance().consumeFail(logGroup.getLogs());
                }
            }
        }

        @SuppressWarnings("unused")
        public void stopThred() {
            loop = false;
            synchronized (this) {
                this.notify();
            }
        }

        @SuppressWarnings("unused")
        public boolean isSleeping() {
            return is_sleeping;
        }

        public void wakeUp() {
            synchronized (this) {
                this.notify();
            }
        }

    }

}

File: dy-statistics-sdk/src/main/java/com/dingyue/statistics/log/AndroidLogClient.kt
package com.dingyue.statistics.log

import com.dingyue.statistics.common.PLItemKey
import com.dingyue.statistics.remote.IUploadLogRO
import com.dingyue.statistics.aliyun.AliyunUploadLogROImpl

import java.util.HashMap
import java.util.concurrent.ConcurrentLinkedQueue

object AndroidLogClient {

    // 定义发送队列
    private val send_queue = ConcurrentLinkedQueue<LogGroup>()
    private var senderThread: LogSendThread
    private var uploadLogRO: IUploadLogRO

    init {
        senderThread = LogSendThread()
        senderThread.start()
        uploadLogRO = AliyunUploadLogROImpl()
        //        uploadLogRO = new TestUploadROImpl();
    }

    fun putLog(logList: List<ServerLog>?) {
        try {
            //按照project和logstore分组，所以log中必须包含project和logstore两个key
            val map = HashMap<String, LogGroup>()
            if (logList != null && logList.isNotEmpty()) {
                for (i in logList.indices) {
                    if (!logList[i].content.containsKey("project") || !logList[i].content.containsKey("logstore")) {
                        continue
                    }
                    val project = logList[i].content["project"] as String
                    val logstore = logList[i].content["logstore"] as String
                    val unikey = project + "_" + logstore
                    if (!map.containsKey(unikey)) {
                        val logGroup = LogGroup("", "", project, logstore)
                        map[unikey] = logGroup
                    }
                    map[unikey]?.putLog(logList[i])
                    AppLog.e("ad", logList[i].content["project"] as String)
                }
                if (!map.values.isEmpty()) {
                    for (logGroup in map.values) {
                        send_queue.add(logGroup)
                        senderThread.wakeUp()
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    // 开启一个线程专门用来发送数据
    private class LogSendThread : Thread() {
        private var loop = true
        private var isSleeping = false
        private val obj = java.lang.Object()

        override fun run() {
            while (loop) {
                isSleeping = false
                if (send_queue.isEmpty()) {
                    isSleeping = true
                    synchronized(obj) {
                        try {
                            obj.wait()
                        } catch (e: InterruptedException) {
                        }
                    }
                    continue
                }
                val logGroup = send_queue.poll()
                val project = logGroup.project
                val logstore = logGroup.logstore
                try {
                    if (project != null && logstore != null) {
                        // 根据不同类型(ZN_USER,ZN_APP_EVENT等)  调用不同上传接口
                        when (logstore) {
                            PLItemKey.ZN_APP_EVENT.logstore -> uploadLogRO.postAppEvent(logGroup)
                            PLItemKey.ZN_APP_APPSTORE.logstore -> uploadLogRO.postInstallAppList(logGroup)
                            PLItemKey.ZN_USER.logstore -> uploadLogRO.postUserLog(logGroup)
                            PLItemKey.ZN_PV.logstore -> uploadLogRO.postReadContent(logGroup)
                            PLItemKey.ZN_APP_FEEDBACK.logstore -> uploadLogRO.postFeedback(logGroup)
                        }
                    }
                    // 是否要有重试功能.......
                } catch (e: Exception) {
                    e.printStackTrace()
                    AndroidLogStorage.consumeFail(logGroup.logs)
                }
            }
        }

        fun stopThread() {
            loop = false
            synchronized(obj) {
                obj.notify()
            }
        }

        fun wakeUp() {
            synchronized(obj) {
                obj.notify()
            }
        }
    }

}
